imagebuildah/stage_executor.go

```diff
@@ -966,19 +967,21 @@ func (s *StageExecutor) prepare(ctx context.Context, from string, initializeIBCo
 		}
 	}
 
-	builderSystemContext := s.executor.systemContext
-	// get platform string from stage
-	if stage.Builder.Platform != "" {
-		os, arch, variant, err := parse.Platform(stage.Builder.Platform)
+	builderSystemContext := s.systemContext
+	// In a multi-stage build where `FROM --platform=<>` was used then we must
+	// reset context for new stages so that new stages don't inherit unexpected
+	// `--platform` from prior stages.
+	if stage.Builder.Platform != "" || (len(s.stages) > 1 && ((builderSystemContext.ArchitectureChoice == "" && builderSystemContext.VariantChoice == "" && builderSystemContext.OSChoice == "") || (s.executor.skipUnusedStages == types.OptionalBoolFalse))) {
+		imageOS, imageArch, imageVariant, err := parse.Platform(stage.Builder.Platform)
 		if err != nil {
 			return nil, fmt.Errorf("unable to parse platform %q: %w", stage.Builder.Platform, err)
 		}
-		if arch != "" || variant != "" {
-			builderSystemContext.ArchitectureChoice = arch
-			builderSystemContext.VariantChoice = variant
+		if imageArch != "" || imageVariant != "" {
+			builderSystemContext.ArchitectureChoice = imageArch
+			builderSystemContext.VariantChoice = imageVariant
 		}
-		if os != "" {
-			builderSystemContext.OSChoice = os
+		if imageOS != "" {
+			builderSystemContext.OSChoice = imageOS
```

Comment:
Why are we creating `builderSystemContext`, which is a copy of the pointer value in `s.systemContext`, and modifying it through that pointer?  It has the same effect as modifying `s.systemContext` directly, but is harder to follow.  Is it intended to be a one-off copy that we just pass in the `BuilderOptions`?

Reply:
Removed. This is not required since we are already copying context for each stage in `executor`.